lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py

```diff
@@ -438,6 +456,109 @@ def restart_services(self, services):
         """
         self._magmad_client.restart_services(services)
 
+    def create_mme_config_backup(self):
+        """
+        This function creates a backup of default MME configuration file,
+        which can later be used to restore the original configuration
+        """
+
+        print("Creating a backup of MME configuration file")
+        self.exec_command(
+            "sudo cp -n "
+            + self.mme_config_file
+            + " "
+            + self.mme_config_file
+            + "_bak"
+        )
+
+    def reset_mme_config_file(self):
+        """
+        Reset the MME configuration from the backup configuration file
+        """
+
+        print("Restoring MME configuration from backup configuration file")
+        self.exec_command(
+            "sudo cp " + self.mme_config_file + "_bak " + self.mme_config_file
+        )
+
+    def clear_default_plmn_tac_configuration(self):
+        """
+        Remove default PLMN and TAC from MME configuration file
+        """
+
+        print("Removing default GUMMEI_LIST from MME configuration")
+        self.exec_command(
+            "sudo sed -i '/GUMMEI_LIST/{n;d}' " + self.mme_config_file
+        )
+
+        print("Removing default TAI_LIST from MME configuration")
+        self.exec_command(
+            "sudo sed -i '/TAI_LIST/{n;N;N;N;N;N;N;d}' " + self.mme_config_file
+        )
+
+        print("Removing default TAC_LIST from MME configuration")
+        self.exec_command(
+            "sudo sed -i '/TAC_LIST/{n;N;N;d}' " + self.mme_config_file
+        )
+
+    def configure_multiple_plmn_tac(self):
+        """
+        Configure multiple PLMNs and TACs in MME configuration file
+        """
+        config_list = [
+            {"MCC": "001", "MNC": "01", "TAC": "1"},
+            {"MCC": "001", "MNC": "02", "TAC": "2"},
+            {"MCC": "001", "MNC": "03", "TAC": "3"},
+            {"MCC": "001", "MNC": "04", "TAC": "4"},
+            {"MCC": "001", "MNC": "05", "TAC": "5"},
+        ]
+
+        print("Updating multiple PLMNs and TACs in MME configuration")
+        gummei_cmd_str = ""
+        tac_cmd_str = ""
+        count_tacs = len(config_list)
+        for idx in range(count_tacs):
+            gummei_cmd_str += "\ \ \ \ \ \ \ \ \ "
```

Comment:
I have fixed other comments for this PR. I tried this suggestion of simplifying this logic here but it was making the code more complex and hence I left it like this till your next response. The reason is that the config file (mme.conf.template) is not actually mapped to proper JSON format and MME fails to read the config file if we simply update the MME config file with this converted string of JSON (even after escaping the characters). For example,

This is a JSON format example code,
{"MCC": "001", "MNC": "01", "MME_GID": "1", "MME_CODE": "1"},

And this is a MME config example:
{ MCC="001" ; MNC="01" ; MME_GID="1" ; MME_CODE="1" ; },

As we can see there are differences with quotes in keys, colon/equal to and comma/semi colon. Therefore, directly converting the dictionary to string and then escaping does not help here. It needs some complex steps like following:
1) Using pprint or json.dump to convert JSON to string like this:
stringData = pprint.pformat(gummei_config, indent=8).replace("'", "\"")
2) Complex regular expression to unquote the keys only from the string something like this:
stringData = re.sub(r'(?<!: )"(\S*?)"', '\\1', stringData)
3) Replace the brackets with blank as they are already formatted in config file and replace semi-colon with equal to like this:
stringData = replace('(', ' ').replace(')', ' ') .replace(':', '=')      or may be using regex
4) Replace the comma (inside the braces only and not the outside ones) with semicolon using regex
5) Escape the string and then match new lines properly with sed pattern (because it is not same as converted string)

I could not find a better and simpler approach in this suggested method. Please update if you are aware of a simpler approach here which I am missing.

Reply:
Updated the PR with the suggested change.